hystrix-core/src/main/java/com/netflix/hystrix/collapser/CollapsedRequestObservableFunction.java

```diff
@@ -143,118 +144,14 @@ public Exception setExceptionIfResponseNotReceived(Exception e, String exception
      */
     @Override
     public void setException(Exception e) {
-        while (true) {
-            CollapsedRequestObservableFunction.ResponseHolder<T> r = rh.get();
-            if (r.getException() != null) {
-                throw new IllegalStateException("setException can only be called once");
-            }
-            if (r.isResponseSet()) {
-                throw new IllegalStateException("Response is already set so exception can not be => Response: " + r.getResponse() + "  Exception: " + e.getMessage(), e);
-            }
-
-            if (subscription.isUnsubscribed()) {
-                return;
-            }
-            ResponseHolder<T> nr = r.setException(e);
-            if (rh.compareAndSet(r, nr)) {
-                // success
-                sendResponseIfRequired(subscription, nr);
-                break;
-            } else {
-                // we'll retry
-            }
+        if (!isTerminal()) {
+            responseSubject.onError(e);
+        } else {
+            throw new IllegalStateException("Response has already terminated so exception can not be set", e);
         }
     }
 
-    @Override
-    public void call(Subscriber<? super T> observer) {
-        observer.add(subscription);
-        while (true) {
-            CollapsedRequestObservableFunction.ResponseHolder<T> r = rh.get();
-            if (r.getObserver() != null) {
-                throw new IllegalStateException("Only 1 Observer can subscribe. Use multicast/publish/cache/etc for multiple subscribers.");
-            }
-            ResponseHolder<T> nr = r.setObserver(observer);
-            if (rh.compareAndSet(r, nr)) {
-                // success
-                sendResponseIfRequired(subscription, nr);
-                break;
-            } else {
-                // we'll retry
-            }
-        }
+    private boolean isTerminal() {
```

Comment:
Perhaps `isTerminated` makes more sense since it is a check for whether it has happened already, not whether the current item is the last?

I had to confirm what `isTerminal` meant by looking at the code.